### DISCUSSION TITLE
Resource locks with `@BeforeEach`/`@AfterEach` lifecycle methods

### DISCUSSION THREAD

**Discussion by Fanael** [junit-team/junit-framework#3666] [Q&A]
Suppose we have the following:
```java
class ExampleTest {
    @BeforeEach
    void before() {
    }

    @AfterEach
    void after() {
    }

    @ResourceLock("example")
    @Test
    void test() {
    }
}
```

What, exactly, are the guaranteed semantics of acquisition and release of the `example` resource lock when the `before` and `after` methods are invoked for `test`? Is it:
 - acquired once, held across the three methods, then released? 
 - acquired before calling each of the three method and released after each return?
 - only acquired for the duration of `test`, potentially allowing the code in `before` and `after` to race?
 - completely unspecified, so nobody should rely on any particular behavior?

The [SharedResourcesDemo](https://github.com/junit-team/junit5/blob/46d0f80db0d6fc5faced28e9827683a09e7f8fb9/documentation/src/test/java/example/SharedResourcesDemo.java) example from the guide seems to indicate the first interpretation is correct, because I *think* it's possible to show it having undesired behavior under other interpretations, but I'd rather be 100% sure.

**Comment by sbrannen** [comment]
- Closed in favor of #3671
